process/block/preprocess/transactions.go

```diff
@@ -220,65 +232,249 @@ func (txs *transactions) ProcessBlockTransactions(
+				process.ErrInvalidMiniBlockType,
+				miniBlock.Type,
+				miniBlock.SenderShardID,
+				miniBlock.ReceiverShardID)
+		}
+
+		txsFromMiniBlock, err := txs.computeTxsFromMiniBlock(miniBlock)
+		if err != nil {
+			return nil, err
+		}
+
+		allTxs = append(allTxs, txsFromMiniBlock...)
+	}
+
+	return allTxs, nil
+}
+
+func (txs *transactions) computeTxsFromMe(body block.Body) ([]*txcache.WrappedTransaction, error) {
+
+	allTxs := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for _, miniBlock := range block.MiniBlockSlice(body) {
+		shouldSkipMiniblock := miniBlock.SenderShardID != txs.shardCoordinator.SelfId() || !txs.isMiniBlockCorrect(miniBlock.Type)
+		if shouldSkipMiniblock {
 			continue
 		}
 
-		gasConsumedByMiniBlockInSenderShard := uint64(0)
-		gasConsumedByMiniBlockInReceiverShard := uint64(0)
+		txsFromMiniBlock, err := txs.computeTxsFromMiniBlock(miniBlock)
+		if err != nil {
+			return nil, err
+		}
 
-		for j := 0; j < len(miniBlock.TxHashes); j++ {
-			if !haveTime() {
-				return process.ErrTimeIsOut
-			}
+		allTxs = append(allTxs, txsFromMiniBlock...)
+	}
+
+	return allTxs, nil
+}
+
+func (txs *transactions) computeTxsFromMiniBlock(miniBlock *block.MiniBlock) ([]*txcache.WrappedTransaction, error) {
+	txsFromMiniBlock := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for i := 0; i < len(miniBlock.TxHashes); i++ {
+		txHash := miniBlock.TxHashes[i]
+		txs.txsForCurrBlock.mutTxsForBlock.RLock()
+		txInfoFromMap := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+		txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
+
+		if txInfoFromMap == nil || txInfoFromMap.tx == nil {
+			log.Debug("missing transaction in computeTxsFromMiniBlock", "type", miniBlock.Type, "txHash", txHash)
+			return nil, process.ErrMissingTransaction
+		}
+
+		tx, ok := txInfoFromMap.tx.(*transaction.Transaction)
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		calculatedSenderShardId, err := txs.getShardFromAddress(tx.GetSndAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		calculatedReceiverShardId, err := txs.getShardFromAddress(tx.GetRecvAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		wrappedTx := &txcache.WrappedTransaction{
+			Tx:              tx,
+			TxHash:          txHash,
+			SenderShardID:   calculatedSenderShardId,
+			ReceiverShardID: calculatedReceiverShardId,
+		}
+
+		txsFromMiniBlock = append(txsFromMiniBlock, wrappedTx)
+	}
+
+	return txsFromMiniBlock, nil
+}
+
+func (txs *transactions) getShardFromAddress(address []byte) (uint32, error) {
+	isEmptyAddress := bytes.Equal(address, make([]byte, txs.addressConverter.AddressLen()))
+	if isEmptyAddress {
+		return txs.shardCoordinator.SelfId(), nil
+	}
+
+	addressContainer, err := txs.addressConverter.CreateAddressFromPublicKeyBytes(address)
+	if err != nil {
+		return 0, err
+	}
+
+	return txs.shardCoordinator.ComputeId(addressContainer), nil
```

Comment:
why don't we make a method at shardCoordinator to computeId from byte array ?

Reply:
I think this should be done in a separate PR, due the number of lines which should be modified in mocks and stubs.

Reply:
It is not so complicated. But a junior can do it.

Reply:
The function is already implemented in shardCoordinator, it have to be used :)

Reply:
I have only added in the implementation, but not in the interface and in all mocks just to keep this PR under 3000 lines :). It should be done in a separate short PR as I have already prepared the method